upgrade: wait for API server before continuing

charlesprost opened the pull request:
**Component**:

'containers', 'tests', 'ci'

**Context**: 

If we don't wait for apiserver to be up and ready during the upgrade process, then apiserver may be still booting, then any request to it will fail hence the whole upgrade process will fail.

Thx @TeddyAndrieux for helping on this one.

**Acceptance criteria**: 

Upgrade process should not fail because apiserver is not available.

---

Closes: #1836

charlesprost committed:
upgrade: wait for API server before continuing

If we don't wait, then apiserver may be still booting, then any request
to it will fail hence the whole upgrade process will fail.

Fixes: #1836

bert-e commented:
# Hello nootal,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/1836-wait-for-apiserver-when-upgrading` into
`development/2.1`
* [w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.2` 
* [w/2.3/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.3` 
* [w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

TeddyAndrieux commented on salt/metalk8s/orchestrate/upgrade/init.sls:
```diff
@@ -41,6 +41,13 @@ Skip node {{ node }}, already in {{ node_version }} newer than {{ dest_version }
+  - name: https://{{ pillar.metalk8s.api_server.host }}:6443/healthz
+  - match: 'ok'
+  - status: 200
+  - verify_ssl: false
```
I think you should move the requires here
```
      - salt: Upgrade etcd cluster
    {%- if previous_node is defined %}
      - salt: Deploy node {{ previous_node }}
    {%- endif %}
```

charlesprost replied:
Done.

TeddyAndrieux commented on salt/metalk8s/orchestrate/upgrade/init.sls:
```diff
@@ -41,6 +41,13 @@ Skip node {{ node }}, already in {{ node_version }} newer than {{ dest_version }
 
   {%- else %}
 
+Wait for API server to be available:
```
That will not work as you are in a `for` two state will have the same name you need to add the node name here.
Maybe `Wait for API server to be available for {{ node }}`

NicolasT replied:
`on`, not `for`.

TeddyAndrieux replied:
Hmm it's not really `on` because we not check API server on this node but we check that API server available to be able to upgrade the node

TeddyAndrieux replied:
Oh but yes we wait **on** this node so `on` maybe :stuck_out_tongue:

TeddyAndrieux reviewed (commented):
IMO same thing should be done in the downgrade orchestrate.

And that should target development/2.0 no ?

charlesprost commented:
> IMO same thing should be done in the downgrade orchestrate.

I'll create a ticket for this.

> 
> And that should target development/2.0 no ?

No because in 2.0 we don't have the upgrade test (since we don't test from a version below 2.0).

TeddyAndrieux commented:
> No because in 2.0 we don't have the upgrade test (since we don't test from a version below 2.0).

Yes but your change are in the upgrade orchestrate so part of 2.0. And if one day we go for a 2.0.1 we will include upgrade test also in 2.0. (I don't know if we will do this one day but for me we should also do this in the 2.0 branch)

bert-e commented:
# History mismatch



Merge commit #8bbc48a3af0490f48b981eaf19a7dff8d532524e on the integration branch
`w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading` is merging a branch which is neither the current
branch `bugfix/1836-wait-for-apiserver-when-upgrading` nor the development branch
`development/2.2`.

It is likely due to a rebase of the branch `bugfix/1836-wait-for-apiserver-when-upgrading` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

charlesprost commented:
> > No because in 2.0 we don't have the upgrade test (since we don't test from a version below 2.0).
> 
> Yes but your change are in the upgrade orchestrate so part of 2.0. And if one day we go for a 2.0.1 we will include upgrade test also in 2.0. (I don't know if we will do this one day but for me we should also do this in the 2.0 branch)

You're right, I've rebased and changed the target branch.

charlesprost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/1836-wait-for-apiserver-when-upgrading` into
`development/2.0`
* [w/2.1/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.1` 
* [w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.2` 
* [w/2.3/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.3` 
* [w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

TeddyAndrieux reviewed (dismissed):
LGTM

charlesprost commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/7b5141d192b1b5dc30d5787ebbaedbe8b75150fd) did not succeed in branch w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/4a6743fd63b6b33e8a5ea4c439a33fb3a7eca887) did not succeed in branch w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b1fb5e35209c33f7b9d62827c88d6af661876667) did not succeed in branch bugfix/1836-wait-for-apiserver-when-upgrading.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #b1fb5e35209c33f7b9d62827c88d6af661876667 on the integration branch
`w/2.1/bugfix/1836-wait-for-apiserver-when-upgrading` is merging a branch which is neither the current
branch `bugfix/1836-wait-for-apiserver-when-upgrading` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/1836-wait-for-apiserver-when-upgrading` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

charlesprost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/1836-wait-for-apiserver-when-upgrading` into
`development/2.0`
* [w/2.1/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.1` 
* [w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.2` 
* [w/2.3/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.3` 
* [w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/325c106e43a38287de92de5a0db71efba426baa8) did not succeed in branch bugfix/1836-wait-for-apiserver-when-upgrading.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/f0669bdde193991349ada09e78509c1c72316c0e) did not succeed in branch w/2.4/bugfix/1836-wait-for-apiserver-when-upgrading.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye nootal.
